src/network_event/client_event.rs

```diff
@@ -119,155 +119,219 @@ impl ClientEventAppExt for App {
             .create_client_channel(channel.into());
 
         self.add_event::<T>()
-            .init_resource::<Events<FromClient<T>>>()
-            .insert_resource(ClientEventChannel::<T>::new(channel_id))
-            .add_systems(
-                PreUpdate,
-                (
-                    reset::<T>.in_set(ClientSet::ResetEvents),
-                    receive_system
-                        .in_set(ServerSet::Receive)
-                        .run_if(server_running),
-                ),
-            )
-            .add_systems(
-                PostUpdate,
-                (
-                    send_system.run_if(client_connected),
-                    resend_locally::<T>.run_if(has_authority),
-                )
-                    .chain()
-                    .in_set(ClientSet::Send),
-            );
+            .init_resource::<Events<FromClient<T>>>();
+
+        self.world
+            .get_resource_mut::<ClientEventRegistry>()
+            .unwrap()
+            .register_event::<T>(channel_id, send_fn, receive_fn);
 
         self
     }
 }
 
-fn receive<T: Event + DeserializeOwned>(
-    mut client_events: EventWriter<FromClient<T>>,
-    mut server: ResMut<RepliconServer>,
-    channel: Res<ClientEventChannel<T>>,
-) {
-    for (client_id, message) in server.receive(*channel) {
-        match DefaultOptions::new().deserialize(&message) {
-            Ok(event) => {
-                trace!(
-                    "applying event `{}` from `{client_id:?}`",
-                    any::type_name::<T>()
-                );
-                client_events.send(FromClient { client_id, event });
-            }
-            Err(e) => debug!("unable to deserialize event from {client_id:?}: {e}"),
+fn map_and_send<T: Event + MapEntities + Serialize + Clone>(world: &mut World, channel: u8) {
+    world.resource_scope(|world, mut client: Mut<RepliconClient>| {
+        let entity_map = world.get_resource::<ServerEntityMap>().unwrap();
+        let events = world.get_resource::<Events<T>>().unwrap();
+
+        for mut event in events.get_reader().read(events).cloned() {
+            event.map_entities(&mut EventMapper(entity_map.to_server()));
+            let message = DefaultOptions::new()
+                .serialize(&event)
+                .expect("mapped client event should be serializable");
+
+            trace!("sending event `{}`", any::type_name::<T>());
+            client.send(channel, message);
         }
-    }
+    });
 }
 
-fn send<T: Event + Serialize>(
-    mut events: EventReader<T>,
-    mut client: ResMut<RepliconClient>,
-    channel: Res<ClientEventChannel<T>>,
-) {
-    for event in events.read() {
-        let message = DefaultOptions::new()
-            .serialize(&event)
-            .expect("client event should be serializable");
-
-        trace!("sending event `{}`", any::type_name::<T>());
-        client.send(*channel, message);
-    }
+struct ClientEventFns {
```

Comment:
I think that this struct will be the same for server events too. I suggest to move it to the `NetworkEvent` module and rename into `NetworkEventFns`.